src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -67,12 +67,78 @@ QgsPointCloudLayerRenderer::QgsPointCloudLayerRenderer( QgsPointCloudLayer *laye
   mReadyToCompose = false;
 }
 
+float decodeElevation( const QColor &elevationColor )
+{
+  return elevationColor.redF() + ( elevationColor.greenF() + elevationColor.blueF() / 255.0 ) / 255.0;
+}
+
+void applyEyeDomeLighting( QImage *img, QImage *elevationImg, QgsPointCloudRenderContext &context )
+{
+  double strength = context.eyeDomeLightingStrength();
+  double distance = context.eyeDomeLightingDistance();
+
+  float minZ = context.minZ();
+  float maxZ = context.maxZ();
+
+  for ( int i = distance; i < img->width() - distance; ++i )
+  {
+    for ( int j = distance; j < img->height() - distance; ++j )
+    {
+      QColor originalColor = img->pixelColor( i, j );
```

Comment:
Calling pixelColor for each pixel is extremely inefficient. Best to read the image data directly using scan lines or QImage::constBits. See https://doc.qt.io/qt-5/qimage.html#pixelColor